doc/config-generic-update-rollback/SONiC_Generic_Config_Update_and_Rollback_Design.md

```diff
@@ -0,0 +1,617 @@
+- Executing `sudo sonic-cfggen -j /tmp/dhcp.json --write-to-db`
+- Restart `dhcp_relay` service
+
+We have explored [SONiC CLI commands](https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md) to make configuration changes. These CLI commands result in updates to the ConfigDB which are corresponding to the CLI command executed. For example, the config `vlan add 10` will create a new row in the VLAN table of the ConfigDB. But relying on the CLI commands to do partial update is also not feasable as there is no standard way of showing the config after the update. Setting up a different update mechanism for each part of the config is very time consuming and ineffecient.
+
+The other challenge to updating a switch is recoverability via rollback. Rollback needs to be hitless/non-disruptive e.g. if reverting ACL updates DHCP should not be affected. Currently SONiC has a couple of operations that can be candidates for rollback `config load` and `config reload`.
+
+*config reload <config_db.json>* : This command clears all the contents of the ConfigDB and loads the contents of config_db.json into the ConfigDB. After that all the Docker containers  and Linux services are restarted to establish the user specified configuration state in the config_db.json file.
+
+- Pro's
+  - Assured way of affecting a configuration state change
+- Con's
+  - Brings the links down and resets the forwarding state. This operation is disruptive in nature
+  - Time consuming as it may take 2-3 minutes for all the services to come back online. The time taken may vary based on the switch CPU power.
+- Verdict
+  - Cannot be used as a rollback mechanism
+
+*config load  <config_db.json>*: This command loads the contents of config_db.json into the ConfigDB. The updates made to the ConfigDB are additive in nature and thus the new configuration state is a combination of the current running state and the partial configuration state specified by the user in the config_db.json file
+
+- Pro's
+  - Quick way to add new configuration changes
+  - It does not disrupt existing service whose configuration is not being modified. So it is non-disruptive in nature
+- Con's
+  - Can't remove existing configuration and can only be used to add/modify the existing configuration
+- Verdict
+  - Cannot be used as a rollback mechanism
+
+Since both `config load` and `config reload` are not suitable for a hitless/non-disruptive rollback, we have to look for other approaches.
+
+In this design document, we will be exploring how to standardize the way to do partial updates, how to take checkpoints and finally how to rollback the configurations.
+
+In summary, this is the flow of an update:
+
+![basic-target-design](files/basic-target-design.png)
+
+And the steps would be:
+```
+admin@sonic:~$ checkpoint mycheckpoint
+admin@sonic:~$ echo "config changes to apply to ConfigDb" > some-config-changes.patch
+admin@sonic:~$ apply-patch ./some-config-changes.patch
```

Comment:
>patch [](start = 49, length = 5)

RFC6902 has defined "File extension(s):  .json-patch".
We discussed .patch.json before, but I think the RFC one is better.